File: app/lib/core/storage/message_search_storage.dart
import 'dart:async';

import 'package:path/path.dart' as p;
import 'package:sqflite/sqflite.dart';

import '../../features/chat/models/message_model.dart';

class MessageSearchResult {
  const MessageSearchResult({
    required this.id,
    required this.roomId,
    required this.roomName,
    required this.senderId,
    required this.senderName,
    required this.content,
    required this.messageType,
    required this.timestampMs,
    required this.relevanceScore,
    this.matchedText,
  });

  final String id;
  final String roomId;
  final String roomName;
  final String senderId;
  final String senderName;
  final String content;
  final String messageType;
  final int timestampMs;
  final double relevanceScore;
  final String? matchedText;

  DateTime get timestamp => DateTime.fromMillisecondsSinceEpoch(timestampMs);
}

class MessageSearchStats {
  const MessageSearchStats({
    required this.totalResults,
    required this.searchTimeMs,
    required this.query,
  });

  final int totalResults;
  final int searchTimeMs;
  final String query;
}

class MessageSearchResponse {
  const MessageSearchResponse({
    required this.results,
    required this.stats,
    required this.hasMore,
  });

  final List<MessageSearchResult> results;
  final MessageSearchStats stats;
  final bool hasMore;
}

/// 本地消息搜索索引（SQLite FTS5）
///
/// 设计目标：
/// - 对齐 desktop 端的本地索引搜索体验：先出本地结果，再可融合服务端搜索结果
/// - 索引字段：room_name / sender_name / content
class MessageSearchStorage {
  const MessageSearchStorage();

  static const _databaseName = 'message_search.db';
  static const _table = 'message_search';
  static const _defaultLimit = 50;

  static Future<Database>? _databaseFuture;

  static Future<Database> _openDatabase() {
    final existing = _databaseFuture;
    if (existing != null) {
      return existing;
    }
    final future = _initDatabase();
    _databaseFuture = future;
    return future;
  }

  static Future<Database> _initDatabase() async {
    final databasesPath = await getDatabasesPath();
    final path = p.join(databasesPath, _databaseName);

    return openDatabase(
      path,
      version: 1,
      onCreate: (db, version) async {
        await db.execute('''
          CREATE VIRTUAL TABLE IF NOT EXISTS $_table USING fts5(
            id UNINDEXED,
            room_id UNINDEXED,
            room_name,
            sender_id UNINDEXED,
            sender_name,
            content,
            message_type UNINDEXED,
            timestamp UNINDEXED
          );
        ''');
      },
    );
  }

  String _normalizeContent(Message message) {
    final trimmed = message.content.trim();
    if (trimmed.isNotEmpty) {
      return trimmed;
    }
    switch (message.type) {
      case MessageType.image:
        return '[图片]';
      case MessageType.audio:
        return '[语音]';
      case MessageType.video:
        return '[视频]';
      case MessageType.file:
        return '[文件]';
      case MessageType.system:
        return '[系统消息]';
      case MessageType.mixed:
        return '[多媒体消息]';
      case MessageType.text:
        return '[消息]';
    }
  }

  /// 替换某个房间的索引（用于保证与本地缓存一致，避免 FTS 表重复记录）
  Future<void> replaceRoomIndex({
    required String roomId,
    required String roomName,
    required List<Message> messages,
    int maxMessages = 200,
  }) async {
    if (roomId.isEmpty) return;

    final db = await _openDatabase();
    final trimmed = messages.length > maxMessages
        ? messages.sublist(messages.length - maxMessages)
        : messages;

    await db.transaction((txn) async {
      await txn.delete(_table, where: 'room_id = ?', whereArgs: [roomId]);

      for (final message in trimmed) {
        if (message.id.isEmpty) continue;
        if (message.isDeleted) continue;

        await txn.insert(_table, {
          'id': message.id,
          'room_id': roomId,
          'room_name': roomName,
          'sender_id': message.senderId,
          'sender_name': message.displaySenderName,
          'content': _normalizeContent(message),
          'message_type': message.type.name,
          'timestamp': message.timestamp.millisecondsSinceEpoch,
        }, conflictAlgorithm: ConflictAlgorithm.replace);
      }
    });
  }

  String _processSearchQuery(String query) {
    final trimmed = query.trim();
    if (trimmed.isEmpty) return '';

    // 如果包含高级语法，直接返回（与 desktop 端一致）
    if (trimmed.contains('"') ||
        trimmed.contains('AND') ||
        trimmed.contains('OR') ||
        trimmed.contains('NOT')) {
      return trimmed;
    }

    final words = trimmed
        .split(RegExp(r'\s+'))
        .where((w) => w.isNotEmpty)
        .toList();
    if (words.isEmpty) return '';
    if (words.length == 1) {
      return '"${words.first}"*';
    }
    return words.map((w) => '"$w"*').join(' AND ');
  }

  Future<MessageSearchResponse> searchMessages({
    required String query,
    String? roomId,
    String? senderId,
    String? messageType,
    int? dateFromMs,
    int? dateToMs,
    int limit = _defaultLimit,
    int offset = 0,
  }) async {
    final stopwatch = Stopwatch()..start();

    final normalizedQuery = query.trim();
    if (normalizedQuery.isEmpty) {
      return const MessageSearchResponse(
        results: [],
        stats: MessageSearchStats(totalResults: 0, searchTimeMs: 0, query: ''),
        hasMore: false,
      );
    }

    final ftsQuery = _processSearchQuery(normalizedQuery);
    if (ftsQuery.isEmpty) {
      return MessageSearchResponse(
        results: const [],
        stats: MessageSearchStats(
          totalResults: 0,
          searchTimeMs: 0,
          query: normalizedQuery,
        ),
        hasMore: false,
      );
    }

    final conditions = <String>[];
    final args = <Object?>[];

    conditions.add('$_table MATCH ?');
    args.add(ftsQuery);

    if (roomId != null && roomId.trim().isNotEmpty) {
      conditions.add('room_id = ?');
      args.add(roomId.trim());
    }

    if (senderId != null && senderId.trim().isNotEmpty) {
      conditions.add('sender_id = ?');
      args.add(senderId.trim());
    }

    if (messageType != null && messageType.trim().isNotEmpty) {
      conditions.add('message_type = ?');
      args.add(messageType.trim());
    }

    if (dateFromMs != null) {
      conditions.add('timestamp >= ?');
      args.add(dateFromMs);
    }

    if (dateToMs != null) {
      conditions.add('timestamp <= ?');
      args.add(dateToMs);
    }

    final whereClause = conditions.isEmpty ? '1' : conditions.join(' AND ');
    final db = await _openDatabase();

    final countRows = await db.rawQuery(
      'SELECT COUNT(1) AS total FROM $_table WHERE $whereClause',
      args,
    );
    final total =
        (countRows.isNotEmpty ? countRows.first['total'] : 0) as int? ?? 0;

    final queryArgs = <Object?>[
      ...args,
      limit.clamp(1, 100),
      offset < 0 ? 0 : offset,
    ];
    final rows = await db.rawQuery('''
        SELECT
          id,
          room_id,
          room_name,
          sender_id,
          sender_name,
          snippet($_table, 5, '<mark>', '</mark>', '...', 20) AS matched_text,
          content,
          message_type,
          timestamp,
          bm25($_table) AS relevance_score
        FROM $_table
        WHERE $whereClause
        ORDER BY relevance_score, timestamp DESC
        LIMIT ? OFFSET ?
      ''', queryArgs);

    final results = <MessageSearchResult>[];
    for (final row in rows) {
      results.add(
        MessageSearchResult(
          id: (row['id'] ?? '').toString(),
          roomId: (row['room_id'] ?? '').toString(),
          roomName: (row['room_name'] ?? '').toString(),
          senderId: (row['sender_id'] ?? '').toString(),
          senderName: (row['sender_name'] ?? '').toString(),
          matchedText: row['matched_text']?.toString(),
          content: (row['content'] ?? '').toString(),
          messageType: (row['message_type'] ?? 'text').toString(),
          timestampMs: (row['timestamp'] as int?) ?? 0,
          relevanceScore: (row['relevance_score'] as num?)?.toDouble() ?? 0,
        ),
      );
    }

    stopwatch.stop();
    final hasMore = (offset + results.length) < total;

    return MessageSearchResponse(
      results: results,
      stats: MessageSearchStats(
        totalResults: total,
        searchTimeMs: stopwatch.elapsedMilliseconds,
        query: normalizedQuery,
      ),
      hasMore: hasMore,
    );
  }
}

File: h5-app/src/storage/message-search-storage.ts
import type { ChatMessage, MessageSearchResponse, MessageSearchResult, MessageType } from '@/types/chat';

import { getLocalDatabase } from './local-database';
import type { SqlAdapter } from './sql-adapter';

interface SearchRow {
  id: string;
  room_id: string;
  room_name: string;
  sender_id: string;
  sender_name: string;
  content: string;
  message_type: string;
  timestamp: number;
  matched_text?: string | null;
  relevance_score?: number | null;
}

type SearchMode = 'fts' | 'like';

interface ReadySearchStorage {
  db: SqlAdapter;
  mode: SearchMode;
}

export class MessageSearchStorage {
  private initializedAdapter: SqlAdapter | null = null;
  private initializedMode: SearchMode | null = null;

  constructor(private readonly adapterFactory: () => Promise<SqlAdapter> = getLocalDatabase) {}

  async replaceRoomIndex(params: {
    roomId: string;
    roomName: string;
    messages: ChatMessage[];
    maxMessages?: number;
  }): Promise<void> {
    const { roomId, roomName, messages, maxMessages = 200 } = params;
    if (!roomId) return;

    const { db } = await this.ready();
    const trimmed = messages
      .filter((message) => message.id && !message.isDeleted)
      .sort((a, b) => a.timestamp - b.timestamp)
      .slice(-maxMessages);

    await db.transaction(async (tx) => {
      await tx.execute('DELETE FROM message_search WHERE room_id = ?', [roomId]);
      for (const message of trimmed) {
        await tx.execute(
          `
            INSERT OR REPLACE INTO message_search
              (id, room_id, room_name, sender_id, sender_name, content, message_type, timestamp)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?)
          `,
          [
            message.id,
            roomId,
            roomName,
            message.senderId,
            message.senderName,
            normalizeContent(message),
            message.type,
            message.timestamp,
          ],
        );
      }
    });
  }

  async searchMessages(params: {
    query: string;
    roomId?: string;
    messageType?: MessageType | string;
    limit?: number;
    offset?: number;
  }): Promise<MessageSearchResponse> {
    const startedAt = performance.now();
    const query = params.query.trim();
    if (!query) {
      return emptyResponse(query);
    }

    const { db, mode } = await this.ready();
    const limit = Math.min(Math.max(params.limit ?? 50, 1), 100);
    const offset = Math.max(params.offset ?? 0, 0);
    const { whereClause, args } = buildSearchWhere(mode, query, params.roomId, params.messageType);

    const countRows = await db.query<{ total: number }>(
      `SELECT COUNT(1) AS total FROM message_search WHERE ${whereClause}`,
      args,
    );
    const rows = await db.query<SearchRow>(
      `
        SELECT
          id,
          room_id,
          room_name,
          sender_id,
          sender_name,
          content,
          message_type,
          timestamp,
          content AS matched_text,
          0 AS relevance_score
        FROM message_search
        WHERE ${whereClause}
        ORDER BY timestamp DESC
        LIMIT ? OFFSET ?
      `,
      [...args, limit, offset],
    );

    const total = Number(countRows[0]?.total ?? rows.length);
    const results = rows.map(toSearchResult);
    return {
      results,
      stats: {
        totalResults: total,
        searchTimeMs: Math.round(performance.now() - startedAt),
        query,
      },
      hasMore: offset + results.length < total,
    };
  }

  private async ready() {
    return this.initialize();
  }

  private async initialize(): Promise<ReadySearchStorage> {
    const db = await this.adapterFactory();
    if (this.initializedAdapter === db && this.initializedMode) {
      return { db, mode: this.initializedMode };
    }
    try {
      await db.execute(`
        CREATE VIRTUAL TABLE IF NOT EXISTS message_search USING fts5(
          id UNINDEXED,
          room_id UNINDEXED,
          room_name,
          sender_id UNINDEXED,
          sender_name,
          content,
          message_type UNINDEXED,
          timestamp UNINDEXED
        )
      `);
      await assertFtsSearchUsable(db);
      this.initializedAdapter = db;
      this.initializedMode = 'fts';
      return { db, mode: 'fts' };
    } catch (error) {
      console.warn('[h5-app] FTS5 消息搜索不可用，降级到 LIKE 查询', error);
      await db.execute(`
        CREATE TABLE IF NOT EXISTS message_search (
          id TEXT PRIMARY KEY,
          room_id TEXT NOT NULL,
          room_name TEXT NOT NULL,
          sender_id TEXT NOT NULL,
          sender_name TEXT NOT NULL,
          content TEXT NOT NULL,
          message_type TEXT NOT NULL,
          timestamp INTEGER NOT NULL
        )
      `);
      await db.execute('CREATE INDEX IF NOT EXISTS idx_message_search_room ON message_search (room_id)');
      await db.execute('CREATE INDEX IF NOT EXISTS idx_message_search_type ON message_search (message_type)');
      this.initializedAdapter = db;
      this.initializedMode = 'like';
      return { db, mode: 'like' };
    }
  }
}

const assertFtsSearchUsable = async (db: SqlAdapter) => {
  await db.query(
    'SELECT id FROM message_search WHERE message_search MATCH ? LIMIT 1',
    ['"__redcode_noop__"'],
  );
};

const normalizeContent = (message: ChatMessage) => {
  const trimmed = message.content.trim();
  if (trimmed) return trimmed;
  const labels: Record<MessageType, string> = {
    text: '[消息]',
    image: '[图片]',
    audio: '[语音]',
    video: '[视频]',
    file: '[文件]',
    system: '[系统消息]',
    mixed: '[多媒体消息]',
  };
  return labels[message.type] ?? '[消息]';
};

const processSearchQuery = (query: string) => {
  const trimmed = query.trim();
  if (!trimmed) return '';
  if (trimmed.includes('"') || /\b(AND|OR|NOT)\b/.test(trimmed)) {
    return trimmed;
  }
  const words = trimmed.split(/\s+/).filter(Boolean);
  if (words.length === 1) return `"${words[0]}"*`;
  return words.map((word) => `"${word}"*`).join(' AND ');
};

const buildSearchWhere = (
  mode: SearchMode,
  query: string,
  roomId?: string,
  messageType?: MessageType | string,
) => {
  const where: string[] = [];
  const args: (string | number)[] = [];
  if (mode === 'fts') {
    where.push('message_search MATCH ?');
    args.push(processSearchQuery(query));
  } else {
    where.push('(room_name LIKE ? ESCAPE ? OR sender_name LIKE ? ESCAPE ? OR content LIKE ? ESCAPE ?)');
    const likeQuery = `%${escapeLikeQuery(query)}%`;
    args.push(likeQuery, '\\', likeQuery, '\\', likeQuery, '\\');
  }
  if (roomId) {
    where.push('room_id = ?');
    args.push(roomId);
  }
  if (messageType) {
    where.push('message_type = ?');
    args.push(messageType);
  }
  return {
    whereClause: where.join(' AND '),
    args,
  };
};

const escapeLikeQuery = (query: string) => query.replace(/[\\%_]/g, (match) => `\\${match}`);

const toSearchResult = (row: SearchRow): MessageSearchResult => ({
  id: row.id,
  roomId: row.room_id,
  roomName: row.room_name,
  senderId: row.sender_id,
  senderName: row.sender_name,
  content: row.content,
  messageType: (row.message_type || 'text') as MessageType,
  timestamp: Number(row.timestamp),
  relevanceScore: Number(row.relevance_score ?? 0),
  matchedText: row.matched_text ?? undefined,
});

const emptyResponse = (query: string): MessageSearchResponse => ({
  results: [],
  stats: {
    totalResults: 0,
    searchTimeMs: 0,
    query,
  },
  hasMore: false,
});
